db/db_impl_open.cc

```diff
@@ -1106,6 +1106,63 @@ Status DB::Open(const DBOptions& db_options, const std::string& dbname,
                       !kSeqPerBatch, kBatchPerTxn);
 }
 
+Status DBImpl::CreateWAL(DBImpl* impl, const EnvOptions& env_options,
+                         uint64_t log_file_num, uint64_t recycle_log_number,
+                         bool create_new_log, bool called_from_open,
+                         const size_t preallocate_block_size,
+                         log::Writer** new_log) {
+  assert(impl != nullptr);
+  Status s;
+  std::unique_ptr<WritableFile> lfile;
+
+  DBOptions db_options =
+      BuildDBOptions(impl->immutable_db_options_, impl->mutable_db_options_);
+  EnvOptions opt_env_options =
+      impl->env_->OptimizeForLogWrite(env_options, db_options);
+  std::string log_fname =
+      LogFileName(impl->immutable_db_options_.wal_dir, log_file_num);
+
+  if (create_new_log) {
+    if (recycle_log_number) {
+      ROCKS_LOG_INFO(impl->immutable_db_options_.info_log,
+                     "reusing log %" PRIu64 " from recycle list\n",
+                     recycle_log_number);
+      std::string old_log_fname =
+          LogFileName(impl->immutable_db_options_.wal_dir, recycle_log_number);
+      s = impl->env_->ReuseWritableFile(log_fname, old_log_fname, &lfile,
+                                        opt_env_options);
+    } else {
+      if (called_from_open) {
+        s = NewWritableFile(impl->immutable_db_options_.env, log_fname, &lfile,
+                            opt_env_options);
+      } else {
+        s = NewWritableFile(impl->env_, log_fname, &lfile, opt_env_options);
+      }
+    }
+
+    if (s.ok()) {
+      lfile->SetWriteLifeTimeHint(impl->CalculateWALWriteHint());
+      lfile->SetPreallocationBlockSize(preallocate_block_size);
+
+      const auto& listeners = impl->immutable_db_options_.listeners;
+      std::unique_ptr<WritableFileWriter> file_writer(
+          new WritableFileWriter(std::move(lfile), log_fname, opt_env_options,
+                                 impl->env_, nullptr /* stats */, listeners));
+      *new_log =
+          new log::Writer(std::move(file_writer), log_file_num,
+                          impl->immutable_db_options_.recycle_log_file_num > 0,
+                          impl->immutable_db_options_.manual_wal_flush);
+
+      if (called_from_open) {
```

Comment:
If the code is only executed for open case, can we move out of this function?

Reply:
Will do.